Dockerfile-base

```diff
@@ -0,0 +1,51 @@
+# builds the base image of rv32emu (the toolchains)
+# docker build --progress=plain -t sysprog21/rv32emu-base -f Dockerfile-base . 2>&1 | tee build.log
+FROM ubuntu:22.04
+
+# Install packages required for the emulator to compile and execute correctly
+RUN apt-get update && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y \
+    libsdl2-dev libsdl2-mixer-dev python3-pip git 
+
+RUN python3 -m pip install git+https://github.com/riscv/riscof 
+
+# set up the timezone
+ENV TZ=Asia/Taipei
+RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
+
+# when using apt install gcc-riscv64-unknown-elf, this will cause "unsupported ISA subset 'z'" during compilation
+# thus, we are building from scratch, following the version here -> https://github.com/sysprog21/rv32emu/blob/master/.ci/riscv-toolchain-install.sh
+# for x86, we can optimize this part and take the nightly build directly, but not for aarch64 
+ENV RISCV=/opt/riscv
+ENV PATH=$RISCV/bin:$PATH
+WORKDIR $RISCV
+RUN apt install -y autoconf automake autotools-dev curl libmpc-dev libmpfr-dev libgmp-dev gawk build-essential bison flex texinfo gperf libtool patchutils bc zlib1g-dev libexpat-dev
+RUN git clone --recursive https://github.com/riscv/riscv-gnu-toolchain
+RUN cd riscv-gnu-toolchain && \
```

Comment:
Can we maintain a repository which preserves the prebuilt GNU Toolchain for RISC-V? Therefore, it saves unnecessary build time.

Reply:
2 separate images are currently being created, namely `sysprog21/rv32emu-gcc` and `sysprog21/rv32emu-sail`. 

For `gcc`, we are using the compiler version `tags/2023.10.06`, as using `apt install gcc-riscv64-unknown-elf` will cause "unsupported ISA subset 'z'" during the compilation of the emulator on M1. 

For `sail`, we set up the toolchain to use `sail-0.16`, as the latest version `0.17.x` series will cause compilation errors. 

For the reference emulator, we are using the commit `9547a30bf84572c458476591b569a95f5232c1c7` from `sail-riscv`, since this is the commit of the emulator that is closest to the time when https://github.com/sysprog21/rv32emu/commit/01b00b6f175f57ef39ffd1f4fa6a611891e36df3#diff-3b436c5e32c40ecca4095bdacc1fb69c0759096f86e029238ce34bbe73c6e68f is added.

The eventual `rv32emu` on Dockerhub will be built using the multi-stage feature, where both of the aforementioned images will be used (copying the necessary binaries only), and then install extra packages for the emulator to compile and execute with full capabilities correctly, then packaged into an image.